src/decoder/plugins/VgmstreamDecoderPlugin.cxx

```diff
@@ -0,0 +1,332 @@
+	unsigned major = version >> 24;
+	unsigned minor = version >> 16 & 0xFF;
+	unsigned patch = version & 0xFFFF;
+	FmtDebug(vgmstream_domain, "vgmstream {}.{}.{}", major, minor, patch);
+
+	cfg.ignore_loop = block.GetBlockValue("ignore_loop", false);
+	cfg.force_loop = block.GetBlockValue("force_loop", false);
+	cfg.really_force_loop = block.GetBlockValue("really_force_loop", false);
+	cfg.ignore_fade = block.GetBlockValue("ignore_fade", false);
+
+	// same defaults as plugins shipped with vgmstream
+	VgmstreamGetDouble(block, &cfg.loop_count, "loop_count", 2.0);
+	VgmstreamGetDouble(block, &cfg.fade_time, "fade_time", 10.0);
+	VgmstreamGetDouble(block, &cfg.fade_delay, "fade_delay", 0.0);
+
+	return true;
+}
+
+static AudioFormat
+VgmstreamGetFormat(libvgmstream_t *lib, libstreamfile_t *file)
+{
+	uint32_t sample_rate = lib->format->sample_rate;
+	auto sample_format = SampleFormat::UNDEFINED;
+	uint8_t channels = lib->format->channels;
+
+	if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM16) {
+		sample_format = SampleFormat::S16;
+	} else if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM24) {
+		// mpd doesn't support S24 samples, so reopen the stream as S32
+		const int subsong = lib->format->subsong_index;
+
+		libvgmstream_close_stream(lib);
+
+		libvgmstream_config_t new_cfg = cfg;
+		new_cfg.force_sfmt = LIBVGMSTREAM_SFMT_PCM32;
+		libvgmstream_setup(lib, &new_cfg);
+
+		const int err = libvgmstream_open_stream(lib, file, subsong);
+		assert(err == 0);
+		assert(lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM32);
+
+		sample_format = SampleFormat::S32;
+	} else if (lib->format->sample_format == LIBVGMSTREAM_SFMT_PCM32) {
+		sample_format = SampleFormat::S32;
+	} else if (lib->format->sample_format == LIBVGMSTREAM_SFMT_FLOAT) {
+		sample_format = SampleFormat::FLOAT;
+	}
+
+	return {sample_rate, sample_format, channels};
+}
+
+static int
+ParseSubsongName(const char *base) noexcept
+{
+	base = StringAfterPrefix(base, SUBSONG_PREFIX);
+	if (base == nullptr)
+		return 0;
+
+	char *endptr;
+	const auto track = strtol(base, &endptr, 10);
+	if (endptr == base || *endptr != '.')
+		return 0;
+
+	return static_cast<int>(track);
+}
+
+/**
+ * returns the file path stripped of any /song_xxx.* subsong suffix
+ * and the track number (or 0 if no "song_xxx" suffix is present).
+ */
+static VgmstreamContainerPath
+ParseContainerPath(Path path_fs)
+{
+	const Path base = path_fs.GetBase();
+	int track;
+	if (base.IsNull() || (track = ParseSubsongName(NarrowPath(base))) < 1)
+		return {AllocatedPath(path_fs), 0};
+
+	return {path_fs.GetDirectoryName(), track};
+}
+
+static bool
+VgmstreamScanTags(Path path, libvgmstream_t *lib, TagHandler &handler,
+		  ReplayGainInfo *rgi) noexcept
+{
+	bool found_replay_gain = false;
+	bool found_title = false;
+
+	const auto tags_path = AllocatedPath::Build(path.GetDirectoryName(), "!tags.m3u");
+	libstreamfile_t *tags_file = libstreamfile_open_from_stdio(tags_path.c_str());
```

Comment:
This will break on Windows because `AllocatedPath::c_str()` returns a `const wchar_t*` there.